src/mjolnir/graphbuilder.cc

```diff
@@ -200,27 +201,40 @@ void ConstructEdges(const OSMData& osmdata,
           prev_edge.attributes.way_prior = true;
         }
 
-        if (!edge.attributes.way_begin)
+        // We should add the previous edge now that we know its done
+        if (prev_edge.is_valid())
           edges.push_back(prev_edge);
 
         // Mark the current edge as the next edge one since we processed the first edge
         if (prev_edge.attributes.way_begin)
           edge.attributes.way_next = true;
 
+        // Finish this edge
         prev_edge = edge;
 
-        // Start a new edge if this is not the last node in the way
-        if (current_way_node_index != last_way_node_index) {
+        // Figure out if the way doubled back on itself and if so treat it like the way ended
+        bool doubled_back = false;
+        OSMWayNode next_way_node;
+        while (current_way_node_index != last_way_node_index && way_node.node.duplicate_ &&
+               (next_way_node = *way_nodes[current_way_node_index + 1]).node.duplicate_) {
+          way_node = next_way_node;
+          ++current_way_node_index;
+          doubled_back = current_way_node_index != last_way_node_index;
+        }
+
+        // Either we were done making edges from this way or there is an internal part that is doubled
+        // backed over itself and we need to skip it
+        if (current_way_node_index == last_way_node_index || doubled_back) {
+          edges.push_back(prev_edge);
+          current_way_node_index += !doubled_back;
```

Comment:
this is the trick. if we doubled back we want to treat it like we are processing a new way where the doubling back ends (ie splits off again). in the case that we are doubling back the index is already fast forwarded (from the loop above) to the right one so we dont need to increment in that case. in the other case where we want the next way we do need to increment.